src/ring_theory/integral_closure.lean

```diff
@@ -0,0 +1,280 @@
+/-
+Copyright (c) 2019 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+
+Integral closure of a subring.
+-/
+
+import ring_theory.adjoin
+
+universes u v
+
+open polynomial submodule
+
+variables (R : Type u) {A : Type v}
+variables [comm_ring R] [comm_ring A]
+variables [decidable_eq R] [decidable_eq A]
+variables [algebra R A]
+
+def is_integral (x : A) : Prop :=
+∃ p : polynomial R, monic p ∧ aeval R A x p = 0
```

Comment:
I'm a bit concerned about having needing a module instance for definitions like this, where there is no mention of `smul` in the definition. It could potentially become annoying to have to provide module instances that are never used all of the time.

Reply:
The alternative is that you have to carry a ring-hom around all the time.

Reply:
But I have a ring hom all the time. Let's leave it as is anyway.

Reply:
Maybe you ought to have an algebra (-;

Reply:
I think you're right